x/programs/rust/wasmlanche-sdk/src/state.rs

```diff
@@ -124,15 +134,31 @@ where
     pub fn delete(&mut self, key: K) -> Result<(), Error> {
         self.cache.remove(&key);
 
-        unsafe { host::delete_bytes(&self.program, &key.into()) }
+        let args = GetAndDeleteArgs {
+            caller: self.program,
+            key: key.into().0,
+        };
+
+        let args_bytes = borsh::to_vec(&args).map_err(|_| StateError::Serialization)?;
+
+        host::delete_bytes(&args_bytes)
     }
 
     /// Apply all pending operations to storage and mark the cache as flushed
     fn flush(&mut self) -> Result<(), Error> {
-        for (key, value) in self.cache.drain() {
-            unsafe {
-                host::put_bytes(&self.program, &key.into(), &value)?;
-            }
+        let args_iter = self
+            .cache
+            .drain()
+            .map(|(key, val)| (key.into(), val))
+            .map(|(key, val)| PutArgs {
+                caller: self.program,
+                key: key.0,
+                bytes: val,
+            })
+            .map(|args| borsh::to_vec(&args).map_err(|_| StateError::Serialization));
+
+        for args in args_iter {
+            host::put_bytes(&args?)?;
```

Comment:
Can we consider calling `.collect()` to fail fast in case of a serialization error ?
It also makes the iteration below a little bit cleaner imo

```rust
    /// Apply all pending operations to storage and mark the cache as flushed
    fn flush(&mut self) -> Result<(), Error> {
        let args_iter: Result<Vec<Vec<u8>>, StateError> = self
            .cache
            .drain()
            .map(|(key, val)| (key.into(), val))
            .map(|(key, val)| PutArgs {
                caller: self.program,
                key: key.0,
                bytes: val,
            })
            .map(|args| borsh::to_vec(&args).map_err(|_| StateError::Serialization))
            .collect();

        for args in args_iter? {
            host::put_bytes(&args)?;
        }

        Ok(())
    }
```

Reply:
Iterators are lazy, if there's a failure in serialization, none of the `map` code will ever get called.